service/resource/configmap/resource.go

```diff
@@ -188,7 +191,65 @@ func (r *Resource) GetDeleteState(ctx context.Context, obj, currentState, desire
 }
 
 func (r *Resource) GetUpdateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, interface{}, interface{}, error) {
-	return nil, nil, nil, nil
+	customObject, err := toCustomObject(obj)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+	currentConfigMaps, err := toConfigMaps(currentState)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+	desiredConfigMaps, err := toConfigMaps(desiredState)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+
+	var configMapsToCreate interface{}
+	{
+		configMapsToCreate, err = r.GetCreateState(ctx, obj, currentState, desiredState)
+		if err != nil {
+			return nil, nil, nil, microerror.Mask(err)
+		}
+	}
+
+	var configMapsToDelete []*apiv1.ConfigMap
+	{
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "finding out which config maps have to be deleted")
+
+		for _, currentConfigMap := range currentConfigMaps {
+			if !containsConfigMap(desiredConfigMaps, currentConfigMap) {
+				configMapsToDelete = append(configMapsToDelete, currentConfigMap)
+			}
+		}
+
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", fmt.Sprintf("found %d config maps that have to be deleted", len(configMapsToDelete)))
+	}
+
+	var configMapsToUpdate []*apiv1.ConfigMap
+	if updateallowedcontext.IsUpdateAllowed(ctx) {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "finding out which config maps have to be updated")
+
+		for _, currentConfigMap := range currentConfigMaps {
+			desiredConfigMap, err := getConfigMapByName(desiredConfigMaps, currentConfigMap.Name)
+			if IsNotFound(err) {
+				continue
+			} else if err != nil {
+				return nil, nil, nil, microerror.Mask(err)
+			}
+
+			if isConfigMapModified(desiredConfigMap, currentConfigMap) {
+				configMapsToUpdate = append(configMapsToUpdate, desiredConfigMap)
+			}
+		}
+
+		if len(configMapsToUpdate) != 0 {
+			updatenecessarycontext.SetUpdateNecessary(ctx)
+		}
```

Comment:
`len(configMapsToUpdate) != 0` can be checked in the `ProcessUpdate` code. I'd be happier to see `updateState` being `nil` as an indicator that the update should be skipped.

Additionally when `len(configMapsToUpdate) == 0` the update naturally won't happen. I don't understand what this thing guards. Maybe there is a usecase I can't foresee?

Reply:
> I'd be happier to see updateState being nil as an indicator that the update should be skipped

Done. 

> Additionally when len(configMapsToUpdate) == 0 the update naturally won't happen.

The "update necessary" information is for inter resource communication. When the config maps change we want to update the deployments. The deployments usually do not change. I do not want the deployments to check for config map changes. I want a communication channel that is abstracted. The "update necessary" context package is this abstracted channel. The deployment can then check of updates are allowed and if updates are necessary and then compute the update state for the deployments accordingly.

Reply:
> The "update necessary" information is for inter resource communication.

Thanks, I understand now, but new (maybe unrelated) question emerged in my head: why we want update deployments when config map changes?

Reply:
At least in KVM we have the cloud configs we use to boot guest clusters in config maps. The config map resource modified in this PR manages this. All state changes done to the TPO will cause a change in the config maps. State changes to the TPO will not necessarily, or almost never, cause a natural state change in the deployment manifests. So the config map resource knows when a guest cluster has to be updated. Only updating the config maps does not update the guest clusters. We have to roll the guest cluster nodes and this is done by updating the deployments.

Reply:
https://github.com/giantswarm/kvm-operator/pull/188/commits/aa94a5db1aeaafb0ea1bf9e8f71bada281094b2c - this is not what I meant. I didn't get that "update necessary" is for inter resource communication.

Sorry for the confusion, but I think checking `len(...)==0` is at least as good as checking `nil` if not better.

Reply:
So what can I do for you?

Reply:
Up to you. You can revert if you think it's right.